src/postw90/nerwann.F90

```diff
@@ -0,0 +1,1203 @@
+    real(kind=dp) :: rdum, spn_nk(num_wann), alpha_sq, beta_sq
+    real(kind=dp) :: binwidth, r_num_elec_per_state
+    integer :: BandIdx, loop_f, min_f, max_f
+    logical :: DoSmearing    
+
+    r_num_elec_per_state = real(num_elec_per_state, kind=dp)
+
+    ! Get spin projections for every band
+    !
+    if (spin_decomp) call spin_get_nk(ws_region, pw90_spin, wannier_data, ws_distance, wigner_seitz, &
+                                      HH_R, SS_R, kpt, real_lattice, spn_nk, &
+                                      mp_grid, num_wann, seedname, stdout)
+
+    binwidth = EnergyArray(2) - EnergyArray(1)
+
+    TDFtot_kz=0.0_dp !Putting initial value zero
+
+    do BandIdx = 1, num_wann
+      if (spin_decomp) then
+        ! Contribution to spin-up DOS of Bloch spinor with component
+        ! (alpha,beta) with respect to the chosen quantization axis
+        alpha_sq = (1.0_dp + spn_nk(BandIdx))/2.0_dp ! |alpha|^2
+        ! Contribution to spin-down DOS
+        beta_sq = 1.0_dp - alpha_sq ! |beta|^2 = 1 - |alpha|^2
+      end if
+
+      ! Do not use an adaptive smearing here, it would require the knowledge of second derivatives
+      ! And typically, when working at not too small temperatures, smearing is not needed
+
+      ! Faster optimization: I precalculate the indices
+      ! Value of the smearing in eV; default = 0 eV, i.e. no smearing
+      smear = pw90_nerwann%tdf_smearing%fixed_width
+      if (smear/binwidth < min_smearing_binwidth_ratio) then
+        min_f = max(nint((eig_k(BandIdx) - EnergyArray(1))/ &
+                         (EnergyArray(size(EnergyArray)) - EnergyArray(1)) &
+                         *real(size(EnergyArray) - 1, kind=dp)) + 1, 1)
+        max_f = min(nint((eig_k(BandIdx) - EnergyArray(1))/ &
+                         (EnergyArray(size(EnergyArray)) - EnergyArray(1)) &
+                         *real(size(EnergyArray) - 1, kind=dp)) + 1, size(EnergyArray))
+        DoSmearing = .false.
+      else
+        min_f = max(nint((eig_k(BandIdx) - smearing_cutoff*smear - EnergyArray(1))/ &
+                         (EnergyArray(size(EnergyArray)) - EnergyArray(1)) &
+                         *real(size(EnergyArray) - 1, kind=dp)) + 1, 1)
+        max_f = min(nint((eig_k(BandIdx) + smearing_cutoff*smear - EnergyArray(1))/ &
+                         (EnergyArray(size(EnergyArray)) - EnergyArray(1)) &
+                         *real(size(EnergyArray) - 1, kind=dp)) + 1, size(EnergyArray))
+        DoSmearing = .true.
+      end if
+
+      do loop_f = min_f, max_f
+        if (DoSmearing) then
+          arg = (EnergyArray(loop_f) - eig_k(BandIdx))/smear
+          rdum = utility_w0gauss(arg, pw90_nerwann%tdf_smearing%type_index, stdout, seedname)/smear
+        else
+          rdum = 1._dp/(EnergyArray(2) - EnergyArray(1))
+        end if
+
+!  write(*,6)'loop,rdum,occ,vel=',loop_f,rdum,r_num_elec_per_state,vel(BandIdx,1:3) !Added by Dr.Keivan main write
+!Total TDF 1st+2nd along Bz
```

Comment:
The following lines will become more readable if you pre-define some constants and then use them, like 

```
rdum1 = rdum*(pw90_nerwann%relax_time**(-1))*r_num_elec_per_state*& 
                        physics%elem_charge_SI**3/physics%hbar_SI**2*1.e-5_dp
rdum2 = rdum* r_num_elec_per_state*physics%elem_charge_SI**4/ & 
                                physics%hbar_SI**3*1.e-40_dp
```

Then the next line reads just 
```
TDFtot_kz(XX, loop_f, 1) = TDFtot_kz(XX, loop_f,1)+rdum1*vel_k(BandIdx,1)*vel_k(BandIdx,1)+&
                  rdum2*vel_k(BandIdx,1)*omg_bnd1(BandIdx)
```

Further,  combining the 3 arrays `omg_bnd1` ,`omg_bnd2`, `omg_bnd3` into one 2d-adday, the next 9 lines can be rewritten as a double-nested loop. 

Same aapplies till line 1192